logger.go

```
@@ -0,0 +1,96 @@
+*/
+import "C"
+import (
+	"log"
+	"math"
+	"os"
+	"sync"
+	"unsafe"
+)
+
+var (
+	gCallbackID  C.qdb_log_callback_id
+	gLoggerMutex sync.Mutex
+	gLogger      *log.Logger = nil
+	gLogFilePath string      = ""
+)
+
+func getLogLevel(log_level C.qdb_log_level_t) string {
+	switch log_level {
+	case C.qdb_log_detailed:
+		return "[detail]"
+	case C.qdb_log_debug:
+		return "[debug]"
+	case C.qdb_log_info:
+		return "[info]"
+	case C.qdb_log_warning:
+		return "[warning]"
+	case C.qdb_log_error:
+		return "[error]"
+	case C.qdb_log_panic:
+		return "[panic]"
+	}
+	return ""
+}
+
+func convertDate(d *C.ulong, length int) []C.ulong {
+	date := make([]C.ulong, 6)
+	var temp C.ulong
+	// See https://github.com/mattn/go-sqlite3/issues/238 for details.
+	slice := (*[math.MaxInt32 - 1/unsafe.Sizeof(temp)]C.ulong)(unsafe.Pointer(d))[:length:length]
+	for i, s := range slice {
+		date[i] = s
+	}
+	return date
+}
+
+//export go_callback_log
+func go_callback_log(log_level C.qdb_log_level_t, d *C.ulong, pid C.ulong, tid C.ulong, message_buffer *C.char, message_size C.size_t) {
+	date := convertDate(d, 6)
+	msg := C.GoStringN(message_buffer, C.int(message_size))
+	gLoggerMutex.Lock()
+	defer gLoggerMutex.Unlock()
+	gLogger.Printf("%d-%02d-%02dT%d:%d:%d %-10s %d\t%d\t%s\n", date[0], date[1], date[2], date[3], date[4], date[5], getLogLevel(log_level), pid, tid, msg)
+}
+
+func initLogger(filePath string) {
+	gLoggerMutex.Lock()
+	defer gLoggerMutex.Unlock()
+	gLogFilePath = filePath
+	if gLogger == nil {
+		gLogger = log.New(os.Stdout, "[qdb api] ", 0)
+	}
+	if gLogFilePath != "" {
+		f, err := os.OpenFile(gLogFilePath, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+		if err != nil {
+			gLogger.Printf("Warning: cannot create log file at location '%s', logging to console.\n", gLogFilePath)
+		} else {
+			gLogger.SetOutput(f)
+		}
+	}
+}
+
+func swapCallback() {
+	initLogger(gLogFilePath)
+	err := C.qdb_log_remove_callback(gCallbackID)
+	if err != 0 {
+	}
```

?? If you want to ignore the error, then please add a comment.